### PR TITLE
[main] bug 640925 - Enhance draft and proforma invoice report layouts and upgrade tag definitions

### PR DESCRIPTION
**Pull request by DavidHolusa**
## What & why

Updates the Czech localization (CZL) Draft Invoice and Pro Forma Invoice report layouts and adds an upgrade procedure to automatically set the CZ-specific RDL layouts as default during upgrade. Fixes the VAT calculation logic in the Pro Forma Invoice report extension to use `SalesPost.GetSalesLines` with temporary records (filtering out non-Item line types and calling `UpdateVATOnLines`) instead of directly querying `Sales Line`, which ensures correct VAT amounts. Also adds a `TotalLbl` column to the Pro Forma Invoice layout and updates the RDL template to include new fields and improve alignment.

## Linked work

Fixes [AB#640925](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/640925)

## How I validated this

- [x] I read the full diff and it contains only changes I intended.
- [x] I built the affected app(s) locally with no new analyzer warnings.
- [x] I ran the change in Business Central and confirmed it behaves as expected.
- [x] I added or updated tests for the new behavior, or explained below why none are needed.

**What I tested and the outcome**

- Built the Core Localization Pack for Czech app locally using `Publish.ps1`; build succeeded with no new analyzer warnings.
- Verified the upgraded RDL layout renders correctly for the Standard Sales - Pro Forma Invoice report with all new columns (TotalLbl, VATIdentifier, etc.) properly aligned.
- Confirmed the upgrade codeunit correctly sets the CZ-specific report layout as default when the current default is unmodified (matches metadata default or is empty), and skips when a custom layout is already selected.
- No new tests added — the upgrade logic follows the established pattern used by other layout upgrade procedures in this codeunit, and the RDL/layout changes are visual and validated manually.

## Risk & compatibility

- The upgrade procedure modifies `Report Layout Selection` and `Tenant Report Layout Selection` tables. It only overrides the default layout when no custom selection has been made (the current selection is empty or matches the metadata default), so existing customizations are preserved.
- New `tabledata` permissions added for `"Report Layout Selection"` (im) and `"Tenant Report Layout Selection"` (im) in the upgrade codeunit.
- The VAT calculation change in `CalcVATAmountLinesCZL` filters out non-Item line types, which changes the VAT breakdown on the Pro Forma Invoice report — this is intentional to match the correct behavior.

### CODE REVIEW

**Comment by alexei-dobriansky** [comment]
## Good Sense Reviewer - Round 1

# **Recommendation: Accept with Suggestions**

<details>
<summary><strong>What this PR does</strong></summary>

The change adds Czech RDLC layouts for draft invoice and pro forma invoice, and it sets them as the default layouts during upgrade when no custom default is present. It also rebuilds the pro forma VAT recap from temporary sales lines and updates the RDLC dataset so the printed VAT and total labels match the Czech layout.

The upgrade guard checks the current default before changing it, so a tenant-selected layout is not replaced when it differs from the metadata default. The VAT calculation now uses the same temporary-line path as posting/statistics code and then keeps only item lines, which matches the base pro forma report because the printed line data item is item-only.
</details>

<details>
<summary><strong>Problem-solution fit</strong></summary>

**Fit:** Strong

The requested behavior is clear: use the Czech layouts by default and make the pro forma VAT recap match the lines that the report prints. The diff addresses those parts directly and keeps the upgrade scope limited to the two affected reports.
</details>

<details>
<summary><strong>Suggestions</strong></summary>

**S1** (🟠 Moderate): *Add upgrade coverage for default layouts*
The upgrade writes per-company report layout selections. Please add an automated test for empty/default selection and for an existing custom selection, so future changes do not overwrite tenant choices.

**S2** (🟠 Moderate): *Add VAT recap coverage for pro forma*
The VAT recap now depends on Sales-Post.GetSalesLines and the item-line filter. Please add a dataset-level test with mixed item and non-item lines, so the VAT rows stay aligned with the printed lines.
</details>

<details>
<summary><strong>Risk assessment and necessity</strong></summary>

**Risk:** The upgrade touches Report Layout Selection and Tenant Report Layout Selection for the Czech invoice reports, so the main risk is changing a customer's chosen default layout. The VAT change affects report output only, not posting, but a wrong recap would still show wrong tax information on a customer-facing document.

**Necessity:** The change is needed so the Czech report layouts become active after upgrade and the pro forma invoice shows the right Czech VAT recap. The scope is focused on the affected layouts and the matching report extension logic.
</details>

---
> [AI-PR-REVIEW] version=1 promptVersion=4 system=github pr=11242 round=1 by=alexei-dobriansky at=2026-09-09T07:42:43.3550879Z lastSha=3e6848bd3d342ee64caea4be359552fe454277f6 reviewKey=f2e46753a68821696ec4e5ecfd2544af6bee6f2d42a796c4fb7b14c65a4342ad suggestions=S1@f96e90b2,S2@b20ca421

File: src/Apps/CZ/CoreLocalizationPack/app/Src/ReportExtensions/StdSalesProFormaInvCZL.ReportExt.al [hunk 1]
```diff
@@ -369,12 +375,16 @@ reportextension 11708 "Std. Sales - Pro Forma Inv CZL" extends "Standard Sales -
...
     var
-        SalesLine: Record "Sales Line";
+        TempSalesLine: Record "Sales Line" temporary;
+        SalesPost: Codeunit "Sales-Post";
     begin
         VATAmountLineCZL.DeleteAll();
-        SalesLine.SetRange("Document Type", SalesHeader."Document Type");
-        SalesLine.SetRange("Document No.", SalesHeader."No.");
-        SalesLine.CalcVATAmountLines(0, SalesHeader, SalesLine, VATAmountLineCZL);
+        TempSalesLine.DeleteAll();
```
[the remark this answers is not in the document]
Comment (DavidHolusa) [reply src/Apps/CZ/CoreLocalizationPack/app/Src/ReportExtensions/StdSalesProFormaInvCZL.ReportExt.al]: It's ok

**Comment by alexei-dobriansky** [comment]
## Good Sense Reviewer - Round 2

# **Recommendation: Accept with Suggestions**

<details>
<summary><strong>What this PR does</strong></summary>

The new commit also sets the Czech draft invoice and pro forma invoice layouts during new company initialization. It reuses the same layout lookup and selection helper that the upgrade path uses.

The added install wiring is consistent with the rest of the PR: it handles new companies while the upgrade path handles existing companies. The changed helper visibility is internal to the app, so it does not create a public API break.
</details>

<details>
<summary><strong>Status of previous suggestions</strong></summary>

| ID | Title | Status | Author response |
|---|---|---|---|
| S1 | Add upgrade coverage for default layouts | Not addressed | No automated coverage was added; the new company initialization path now uses the same helper. |
| S2 | Add VAT recap coverage for pro forma | Not addressed | No dataset-level VAT recap coverage was added. |
</details>

<details>
<summary><strong>New observations (commits since round 1)</strong></summary>

*None - the new commit extends the default-layout initialization path and does not introduce a new blocker.*
</details>

<details>
<summary><strong>Risk assessment and necessity</strong></summary>

**Risk:** The layout-selection code touches Report Layout Selection and Tenant Report Layout Selection, so the main risk remains replacing a customer's chosen default layout. The report VAT change affects customer-facing output but not posting; the current build checks are mostly green, with one unrelated-looking legacy test bucket failure still visible.

**Necessity:** The change is needed so upgraded and newly initialized Czech companies use the Czech invoice layouts, and so the pro forma VAT recap matches the printed item lines. The scope remains focused on the affected reports and layout-selection setup.
</details>

---
> [AI-PR-REVIEW] version=1 promptVersion=4 system=github pr=11242 round=2 by=alexei-dobriansky at=2026-09-09T13:43:26Z lastSha=1b4982958f94ce392153adbf32addc4de6516d7b reviewKey=8880f680f93f8ad57c203ee3b2c653259b317c0a631266a52a3dc73f789d720c suggestions=S1@f96e90b2:notaddressed,S2@b20ca421:notaddressed parentRound=1

File: src/Apps/CZ/CoreLocalizationPack/app/Src/Codeunits/InstallApplicationCZL.Codeunit.al [hunk 2]
```diff
@@ -598,6 +600,7 @@ codeunit 11748 "Install Application CZL"
         InitEETServiceSetup();
         CreateSourceCodeSetup();
         ModifyReportSelections();
+        InitDefaultReportLayouts();
```
[the remark this answers is not in the document]
Comment (DavidHolusa) [reply src/Apps/CZ/CoreLocalizationPack/app/Src/Codeunits/InstallApplicationCZL.Codeunit.al:603]: fixed

**Comment by alexei-dobriansky** [comment]
## Good Sense Reviewer - Round 3

# **Recommendation: Accept with Suggestions**

<details>
<summary><strong>What this PR does</strong></summary>

The latest commit reuses the layout customization check before setting the Czech draft invoice and pro forma invoice layouts during new company initialization. That matches the upgrade guard, so both upgraded and newly initialized companies avoid replacing a selected custom layout.

The current net change still sets the Czech layouts when the existing default is empty or still the metadata default, and the pro forma VAT recap is built from the same item-line scope that the base report prints. I did not find a new correctness blocker in the latest changed spans.
</details>

<details>
<summary><strong>Problem-solution fit</strong></summary>

**Fit:** Strong

The requested behavior is clear: use the Czech invoice layouts by default and make the pro forma VAT output match the printed item lines. The diff covers existing companies through upgrade, new companies through initialization, and keeps selected custom layouts out of the overwrite path.
</details>

<details>
<summary><strong>Status of previous suggestions</strong></summary>

| ID | Title | Status | Author response |
|---|---|---|---|
| S1 | Add upgrade coverage for default layouts | Not addressed | No automated coverage was added; the latest commit only applies the same guard to the initialization path. |
| S2 | Add VAT recap coverage for pro forma | Not addressed | No dataset-level VAT recap coverage was added. |
</details>

<details>
<summary><strong>New observations (commits since round 2)</strong></summary>

*None - the latest commit applies the customization guard to initialization and does not add a new blocker.*
</details>

<details>
<summary><strong>Risk assessment and necessity</strong></summary>

**Risk:** The layout-selection code touches Report Layout Selection and Tenant Report Layout Selection, so wrong logic could overwrite a selected customer layout. The latest guard lowers that risk for new company initialization; the VAT change affects customer-facing report output, not posting. Some build checks were still in progress when reviewed.

**Necessity:** The change is needed so upgraded and newly initialized Czech companies use the Czech invoice layouts, and so the pro forma VAT recap matches the printed item lines. The scope stays focused on the affected reports and layout-selection setup.
</details>

---
> [AI-PR-REVIEW] version=1 promptVersion=4 system=github pr=11242 round=3 by=alexei-dobriansky at=2026-09-09T19:05:24Z lastSha=03be5403a20ade14c239dfd6d22344d5cf7d7a9b reviewKey=147b64062281595ea0bb75afb588243a831e280bbf24a17951e82b20aba9f466 suggestions=S1@f96e90b2:notaddressed,S2@b20ca421:notaddressed parentRound=2

**Comment by alexei-dobriansky** [comment]
## Good Sense Reviewer - Round 4

# **Recommendation: Request Changes**

<details>
<summary><strong>What this PR does</strong></summary>

The latest commit changes the pro forma VAT calculation again. It now copies only item sales lines into a temporary buffer and builds the VAT recap from that buffer.

That matches the base report's printed line scope, because the base line data item is filtered to item lines. The full change still adds Czech layouts for draft and pro forma invoices and sets those layouts during upgrade and company initialization. The layout setup path protects some built-in layout changes, but it does not protect default custom layout selections.
</details>

<details>
<summary><strong>Problem-solution fit</strong></summary>

**Fit:** Partial

The requested behavior is clear: use the Czech invoice layouts by default, keep selected custom layouts, and make the pro forma VAT output match the printed item lines. The VAT and layout activation parts are covered, but the current customization check can still overwrite a tenant's default custom layout.
</details>

<details>
<summary><strong>Status of previous suggestions</strong></summary>

| ID | Title | Status | Author response |
|---|---|---|---|
| S1 | Add upgrade coverage for default layouts | Not addressed | No automated upgrade or install coverage was added; the custom-layout case still needs coverage. |
| S2 | Add VAT recap coverage for pro forma | Not addressed | No dataset-level VAT recap coverage was added; the latest commit changed the calculation path again. |
</details>

<details>
<summary><strong>New observations (commits since round 3)</strong></summary>

**S3** (🔴 High): *Custom layout selections can be overwritten*
The customization check only reads `Tenant Report Layout Selection`. A default custom layout is stored on `Report Layout Selection`, so this path treats it as not customized and then clears `"Custom Report Layout Code"`. Check `Report Layout Selection` before calling `SetDefaultReportLayout`, so upgrades and install-per-company reruns keep tenant choices.
</details>

<details>
<summary><strong>Risk assessment and necessity</strong></summary>

**Risk:** The report output change is customer-facing tax information, but it does not change posting. The larger risk is the upgrade/default-layout setup: wrong customization detection can overwrite a tenant's chosen custom layout for the two sales reports.

**Necessity:** The change is needed so Czech companies use the Czech invoice layouts and the pro forma VAT recap matches the printed item lines. The scope is right, but preserving existing custom choices is part of the compatibility requirement and must be fixed before merge.
</details>

---
> [AI-PR-REVIEW] version=1 promptVersion=4 system=github pr=11242 round=4 by=alexei-dobriansky at=2026-09-11T13:22:47Z lastSha=d3c0f8edbcf8f9697e1ab1423cfcc05bacbf605e reviewKey=7c8230cb833ffcb7474429d3c45d4278235b1367bf1645debc8247f246d83ae7 suggestions=S1@f96e90b2:notaddressed,S2@b20ca421:notaddressed,S3@65a4d795:new parentRound=3

**Review by v-janpopr — APPROVED** [review]

**Review by AndersLarsenMicrosoft — APPROVED** [review]
